Basic client support

mattmess1221 opened the pull request:
[**SpongeAPI**](#) | [SpongeForge](https://github.com/SpongePowered/SpongeForge/pull/1445)

I'm going to start with some basic client support right now. Currently there is getting the client world, player, loading a save, joining a world, as well as fetching the integrated server (which can also be done from Game.

mattmess1221 committed:
Remove pause

kashike commented on src/main/java/org/spongepowered/api/util/ThreadContext.java:
```diff
@@ -0,0 +1,39 @@
+public interface ThreadContext {
+
+    /**
+     * Checks if the current thread matches the main thread of the server.
```
Javadocs need updating.

kashike commented on src/main/java/org/spongepowered/api/Sponge.java:
```diff
@@ -174,6 +174,35 @@ public static Server getServer() {
 
     /**
+     * Gets whether a {@link Client} instance is available without throwing an
+     * exception from calling {@link #getServer()}.
```
Javadocs need updating.

mattmess1221 committed:
Fixup javadocs

mattmess1221 committed:
Add api to get server connection

mattmess1221 committed:
Don't return a completable future for loading and joining worlds.
Implementation is hard

mattmess1221 committed:
Add base player for shared code between client and server players

mattmess1221 committed:
Add tab list to server connection

mattmess1221 committed:
Add logical side

mattmess1221 committed:
Add basic client support

# Conflicts:
#	src/main/java/org/spongepowered/api/Game.java

mattmess1221 committed:
licenses

ST-DDT commented on src/main/java/org/spongepowered/api/Client.java:
```diff
@@ -0,0 +1,98 @@
+     *
+     * @return The server connection
+     */
+    Optional<ServerConnection> getServerConnection();
```
Does it support both local and remote servers? If (not) it should be specified.

mattmess1221 replied:
Yes. A `RemoteConnection` can point to a local channel.

ST-DDT replied:
Then please add this to the javadocs so you can be sure about that without knowing anything about the API and just doing some API discovery. (Much info => good documentation)

mattmess1221 committed:
Update javadocs for getServerConnection()

mattmess1221 committed:
Update javadocs for getAssetsDirectory()

mattmess1221 committed:
Add client support for player list

mattmess1221 committed:
Fix licenses

kashike commented on src/main/java/org/spongepowered/api/entity/living/player/Player.java:
```diff
@@ -69,7 +66,7 @@
  * that persists across server restarts.</p>
  */
-public interface Player extends Humanoid, User, RemoteSource, Viewer, ChatTypeMessageReceiver {
+public interface Player extends BasePlayer, RemoteSource, Viewer, ChatTypeMessageReceiver {
```
`ServerPlayer extends Player`

kashike commented on src/main/java/org/spongepowered/api/entity/living/player/BasePlayer.java:
```diff
@@ -0,0 +1,60 @@
+ * Base interface for {@link Player server} and {@link ClientPlayer client}
+ * players.
+ */
+public interface BasePlayer extends Human, User {
```
`Player`

kashike commented on src/main/java/org/spongepowered/api/entity/living/player/tab/PlayerTabList.java:
```diff
@@ -0,0 +1,107 @@
+ * A mutable version of {@link TabList}. It also contains a reference to the
+ * {@link Player} entity.
+ */
+public interface PlayerTabList extends TabList {
```
`ServerTabList`

ST-DDT commented on src/main/java/org/spongepowered/api/Client.java:
```diff
@@ -0,0 +1,100 @@
+     *
+     * @return The current world
+     */
+    Optional<World> getLoadedWorld();
```
I'm not sure about this method.
First of all it is not entirely sure which world this is. Is this the world the client is currently viewing or the main world of the savegame and does it include the local copy of the remote world.

Then the location i would rather move it to the LocalServer interface because it is a substate of that.

liach replied:
the world the client is currently viewing

ST-DDT commented on src/main/java/org/spongepowered/api/Client.java:
```diff
@@ -0,0 +1,100 @@
+     * @param saveLocation The location of the save
+     * @see Game#getSavesDirectory()
+     */
+    void loadWorldSave(Path saveLocation);
```
There should be a way to know whether the loafing succeeded and the server/game is running

ST-DDT replied:
Or note that the method blocks.

ST-DDT commented on src/main/java/org/spongepowered/api/Client.java:
```diff
@@ -0,0 +1,100 @@
+     *
+     * @param server The server to join
+     */
+    void joinServer(InetSocketAddress server);
```
Same here

JBYoshi commented on src/main/java/org/spongepowered/api/entity/living/player/Player.java:
```diff
@@ -24,136 +24,17 @@
-     * @return The event that was thrown from sending the message
-     */
-    MessageChannelEvent.Chat simulateChat(Text message, Cause cause);
+public interface Player extends Human, User {
```
Should be `Humanoid` instead of `Human`.

mattmess1221 committed:
Merge shared client (self) and server player objects into a "controller"

mattmess1221 commented:
Some changes in my latest push:
Earlier, I had completely butchered the Player class. No more of that. Instead, I extracted a couple sub-interfaces. One is a `BasePlayer`, and the other is a `PlayerController`. (I'm not married to those names yet)

Anyway, BasePlayer seems pretty straightforward, so I'll focus on PlayerController.

Using the `PlayerController` and subsequent `ClientPlayerController`, you can control certain things about the player and make it do things. This is basically where I stuck all the shared methods which the client player and all the server players have access to.

The server can open a container on a player, and so can the client to itself. The client is not able to do this to other players.

mattmess1221 committed:
Make the client a ChatTypeMessageReceiever which delegates to the player controller.

mattmess1221 committed:
Add method for sending chat to the server.

mattmess1221 committed:
Rename PlayerTabList to ServerTabList

ST-DDT commented on src/main/java/org/spongepowered/api/Client.java:
```diff
@@ -0,0 +1,111 @@
+    Optional<ClientPlayerController> getPlayerController();
+
+    @Override
+    default void sendMessage(ChatType type, Text message) {
```
Javadocs

ST-DDT commented:
Some default methods have the Override on an newline and some have them in the same line. IMo it should be always in a newline.

mattmess1221 committed:
Update javadocs and move some methods around.

mattmess1221 committed:
Add server connection events.

mattmess1221 committed:
Add client chat events
Includes the entire event chain

Also add a loopback message channel

mattmess1221 committed:
Merge branch 'logicalside' into feature/client

mattmess1221 commented:
Recent changes:
1. Add connection events for the client. Named `ServerConnectionEvent`
2. Deprecated the chat event. It's been replaced by 4 platform specific events. client -> server, server -> client, send, receive.
3. Added logical side to `Extent`
4. Added a loopback `MessageChannel` for either the console or client.

parlough commented on src/main/java/org/spongepowered/api/Game.java:
```diff
@@ -101,6 +101,23 @@
+     * @return The client
+     * @throws IllegalStateException If the Client isn't currently available.
+     */
+    Client getClient();
```
Why not just have this return `Optional<Client>` instead? Errors are a pain and an optional fits well here I think. If it's empty that means there is no client available.

mattmess1221 replied:
I've been keeping it consistent with Server. If someone else agrees, I'll change it.

Zidane replied:
Keep this consistent with getServer, we did it so that everyone's code wouldn't be littered with TONS of getServer().get() calls.

ryantheleach commented on src/main/java/org/spongepowered/api/Client.java:
```diff
@@ -53,6 +56,11 @@
     Optional<ClientPlayerController> getPlayerController();
 
+    @Override
+    default void sendMessage(ChatType type, Text message) {
```
Is this plugin to gui? or client to server?

mattmess1221 replied:
I removed this method earlier. It delegated to the PlayerController, so it would send a message to the client.

mattmess1221 committed:
Remove base and client specific classes. Rebase this commit

mattmess1221 committed:
Merge branch 'bleeding' of https://github.com/SpongePowered/SpongeAPI into feature/client

# Conflicts:
#	src/main/java/org/spongepowered/api/entity/living/player/Player.java
#	src/main/java/org/spongepowered/api/event/network/ClientConnectionEvent.java

Zidane commented on src/main/java/org/spongepowered/api/network/NetworkPlayer.java:
```diff
@@ -0,0 +1,43 @@
+/**
+ * @see TabListEntry
+ */
+public interface NetworkPlayer {
```
Maybe NetworkPlayerEntry? NetworkPlayer itself may imply something different to others.

JBYoshi commented on src/main/java/org/spongepowered/api/entity/living/player/ClientPlayerController.java:
```diff
@@ -0,0 +1,68 @@
+     *
+     * @return The player entity
+     */
+    Humanoid getPlayer();
```
Should this be in `PlayerController` as well? It can be default-implemented in `Player` to `return this`.

mattmess1221 replied:
This actually shouldn't be here. I want to abstractify the player's entity.

JBYoshi commented on src/main/java/org/spongepowered/api/entity/living/player/ClientPlayerController.java:
```diff
@@ -0,0 +1,68 @@
+     * @param message The message to send.
+     * @throws IllegalArgumentException If the message has illegal characters
+     */
+    void sendChatMessage(String message) throws IllegalArgumentException;
```
Should this be part of `PlayerController`?

mattmess1221 replied:
No. If I put it in `PlayerController`, people will see it in `Player` and think it sends a message to the player. In actuality, this method sends a method to the server, which `Player` already has a similar method for. (`simulateChat(Text, Cause)`)

JBYoshi replied:
How about making that the shared method? That name seems to work for both client and server.

JBYoshi commented on src/main/java/org/spongepowered/api/entity/living/player/PlayerController.java:
```diff
@@ -0,0 +1,126 @@
+ *
+ * TODO I don't like the name 'controller'
+ */
+public interface PlayerController extends RemoteSource, ChatTypeMessageReceiver {
```
Not sure if this should extend `RemoteSource` - the `ClientPlayerController` isn't technically "remote".

gabizou commented:
Is this being superseded by https://github.com/SpongePowered/SpongeAPI/pull/1804? Genuine question because I do feel both targeting the same idea, but as I mentioned in the other one, we'll be exposing the client in a more fuller context for API 8 along with some more breakages.

mattmess1221 commented:
I guess it is being superceded. IMO, this request got too big for 1 thing. I would like to focus on my Resources PR in api 8 for now anyway.

Zidane commented:
@killjoy1221 

Would you be interested in re-opening this PR? This was a step in the right direction and I've already merged some client stuff into 1.13. Namely that other PR.

mattmess1221 commented:
Alright, I'll work on it a bit. I will also create a new PR targetting 1.13.
